model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.crl.infinispan;
+
+import java.security.GeneralSecurityException;
+import java.security.cert.X509CRL;
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+import org.infinispan.Cache;
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.crl.CrlStorageProvider;
+
+/**
+ *
+ * @author rmartinc
+ */
+public class InfinispanCrlStorageProvider implements CrlStorageProvider {
+
+    private static final Logger log = Logger.getLogger(InfinispanCrlStorageProvider.class);
+
+    private final Cache<String, X509CRLEntry> crlCache;
+    private final Map<String, FutureTask<X509CRL>> tasksInProgress;
+    private final long cacheTime;
+    private final long minTimeBetweenRequests;
```

Comment:
If you want to reduce memory allocation, create an interface to expose this shared instances. Similar to what I did here

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/remote/RemoteInfinispanClusterProvider.java#L47

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/remote/RemoteInfinispanClusterProvider.java#L161-L166